### DISCUSSION TITLE
F-String Formatting Style

### DISCUSSION THREAD

**Discussion by dhruvmanila** [astral-sh/ruff#9785] [General]
This document provides examples to initiate a discussion on formatting f-strings, particularly the expression part, which can now be broken into multiple lines as of Python 3.12.

The main decision which needs to be taken is when to add line breaks within the replacement field of f-strings, especially when either or all of the following are present:

1. Conversion flag
2. Format specifier
3. Parentheses

Another important one is regarding quotes. Now that we’re allowed to re-use the same quotation style in f-strings, should the same quote be used throughout the f-string?

_Black's implementation: https://github.com/psf/black/pull/3822_
_Ruff's implementation: https://github.com/astral-sh/ruff/pull/9642_

Tagging some folks to initiate a conversation to keep the style compatible with Black @tusharsadhwani @ambv

---

### Examples:

This can be best discussed with various examples:

### 1

A simple expression containing a comment right after the opening curly braces.

```python
x = f"aaaaa {  # comment
    bbbbb } ccccc"
```

Possible formatting style:

```python
x = f"aaaaa {  # comment
    bbbbb
} ccccc"
```

### 2

Taking the example from above and adding a few additional comments:

```python
x = f"aaa {  # comment 1
    # comment 2
    bbb
}"
```

And some more:

```python
x = f"aaa {  # comment 1
    # comment 2
    bbb  # comment 3
    # comment 4
}"  # comment 5
```

The code snippets in (2) illustrate the preferred style for formatting f-strings with comments.

### 3

When there are more than 1 replacement field and the f-string exceeds the line-length

```python
x = f"aaaaaaaaaaaa { bbbbbbbbbbbbbb } cccccccccccccccccccc { ddddddddddddddd } eeeeeeeeeeeeee"
```

Should we only break the expression from left to right?

```python
x = f"aaaaaaaaaaaa {
    bbbbbbbbbbbbbb
} cccccccccccccccccccc { ddddddddddddddd } eeeeeeeeeeeeee"
```

Or, break it from right to left?

```python
x = f"aaaaaaaaaaaa { bbbbbbbbbbbbbb } cccccccccccccccccccc {
	ddddddddddddddd
} eeeeeeeeeeeeee"
```

Or, both expressions?

```python
x = f"aaaaaaaaaaaa {
    bbbbbbbbbbbbbb
} cccccccccccccccccccc {
    ddddddddddddddd
} eeeeeeeeeeeeee"
```

And, should this be parenthesized?

```python
x = (
    f"aaaaaaaaaaaa {
        bbbbbbbbbbbbbb
    } cccccccccccccccccccc {
        ddddddddddddddd
    } eeeeeeeeeeeeee"
)
```

### 4

When the expression is either a dictionary or set, then we need to surround the expression with an additional whitespace to not escape the curly brace. But, this is only to be done if the expression doesn’t fit on the line.

```python
x = f"aaaaa { {'bbb', 'ccc', 'ddd'}} eeeee"
```

Here, we need to preserve the whitespace between the opening curly brace of the replacement field and the set otherwise it behaves as escaping the opening curly braces. **The question is whether to add a whitespace before the closing curly braces for consistency?**

```python
x = f"aaaaa { {'bbb', 'ccc', 'ddd'} } eeeee"
#            ^                     ^ Should we add this?
#            This needs to
#            be preserved
```

### 5

When the expression is a collection exceeding the line-length

```python
x = f"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa { [ 'aaaaaaaaaaaaaaa', 'bbbbbbbbbbbbbbbbbbb', 'ccccccccccccccccccc', 'dddddddddddddddd', 'eeeeeeeeeeeeeee', ] } ddddddddddddd"
```

Should we break it twice? This means to break at both `{` and `[`.

```python
x = f"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa {
    [
        'aaaaaaaaaaaaaaa',
        'bbbbbbbbbbbbbbbbbbb',
        'ccccccccccccccccccc',
        'dddddddddddddddd',
        'eeeeeeeeeeeeeee',
    ]
} ddddddddddddd"
```

Or, hug the parentheses as done in the [preview style](https://github.com/astral-sh/ruff/pull/8293)?

```python
x = f"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa {[
    'aaaaaaaaaaaaaaa',
    'bbbbbbbbbbbbbbbbbbb',
    'ccccccccccccccccccc'
    'dddddddddddddddd',
    'eeeeeeeeeeeeeee',
]} ddddddddddddd"
```

### 6

If the conversion flag is present, should it always be right after the expression or can it go to another line? I think it should be the former, even in the case the expression breaks into multiple lines like the example shown above. So,

```python
x = f"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa { 'bbbbbbbbbbbbbbbbbbbbbbbbbbbbb'!r } ccccccccccccccc"
```

Not breaking before the conversion flag:

```python
x = f"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa {
	'bbbbbbbbbbbbbbbbbbbbbbbbbbbbb'!r
} ccccccccccccccc"
```

But, if there’s a collection instead:

```python
x = f"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa {
	[
        'aaaaaaaaaaaaaaa',
        'bbbbbbbbbbbbbbbbbbb',
        'ccccccccccccccccccc',
    ]!r
} ddddddddddddd"
```

### 7

Should we break before the `:` token if format spec is present and the f-string breaks across multiple lines with expression? This would be a bit similar to formatting a slice expression.

```python
x = f"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa { "bbbbbbbbbbbbbbbbbbbbbbbbbbbbb" + "cccccccccccccc" + "dddddddddddddddddddddd" + "eeeeeeeeeeeeee":>10 } ccccccccccccccc"
```

Should the above be formatted as the following? The downside here is that it’s difficult to notice the format specifier otherwise.

```python
x = f'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa {
    "bbbbbbbbbbbbbbbbbbbbbbbbbbbbb" + "cccccccccccccc" + "dddddddddddddddddddddd" + "eeeeeeeeeeeeee":>10 
} ccccccccccccccc'
```

Or,

```python
x = f'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa {
    "bbbbbbbbbbbbbbbbbbbbbbbbbbbbb" + "cccccccccccccc" + "dddddddddddddddddddddd" + "eeeeeeeeeeeeee"
	:>10 
} ccccccccccccccc'
```

Note that whitespaces aren’t allowed in the format spec. Here, the line break is allowed because the newline character ends the format specifier. This creates an interesting code snippet which is valid during parsing and runtime:

```python
f"{  # comment 1
	foo:>2
	# comment 2
}"
```

Here, the `comment 2` is actually a valid comment. But, if it’s placed right after the format specifier, then it’s invalid and will be part of the format specifier as a string. This is because of https://github.com/astral-sh/ruff/pull/7787.

### 8

Should we re-use the same quotes throughout the f-string?

```python
f"first {f"second" + "third"} fourth"
```

---

### Self-documenting expression

When self-documenting expressions are present in an f-string (`=` after an expression), then all whitespace needs to be preserved which means the formatter shouldn’t take any decisions at all. But, we can still normalize whitespace / format the conversion flag and format spec because that isn’t used in the string value. For example,

```python
print(f"{x   =   !s
         :>0

         }")
```

We can format the above as:

```python
print(f"{x   =   !s:>0}")
```

Because both are equivalent. And, whitespace isn’t allowed in the format specifier otherwise it raises a `ValueError` with the message “Space not allowed in string format specifier”.

Another question here is whether to preserve whitespace for nested f-strings when a self-documenting expression is present in the parent f-string. For example,

```python
>>> x = 1
>>> print(
...     f"""
... {
...     (f"{x    }") =
...  }"""
... )

    (f"{x    }") =
 '1'
```

Here, the whitespace in the nested f-string is preserved in the output (`f”{x    }”`) but if there’s a comment or newline, then we get unexpected output:

```python
>>> x = 1
>>> print(
...     f"""
... {
...     f"{x
... 
... 
... 
...        # comment
...        }" =
...  }"""
... )

    f"{x
 '1'
```

Here, the newlines and comment after the nested f-string (`f”{x`) isn’t preserved in the output.

With that, I think we should stop formatting the current f-string once we encounter a self-documenting expression.

**Comment by Eutropios** [comment]
I think it would be useful to also see how yapf (https://github.com/google/yapf/issues/1136) and autopep8 (https://github.com/hhatto/autopep8/issues/712) decide to implement this formalization.

**Reply by Avasam** [reply]
Speaking of autopep8, I've been tracking how close Ruff is to full parity with its autofixes: https://github.com/astral-sh/ruff/discussions/9057

**Comment by dhruvmanila** [comment]
# Style decision: Indentation

What should be the indentation level of the expression part of an f-string? The problem here is that the formatter can't push the `{` for alignment as otherwise it would mean adding whitespace in the string value of the f-string. But, it can push the `}` as it's all expression before it.

For example:

```python
if indent0:
    if indent1:
        if indent2:
            foo = f"""
# Heading
--------- {
    [
        "aaaaaaaaaaaaaaaaaaaaa",
        "bbbbbbbbbbbbbbbbbbbbb",
        "ccccccccccccccccccccc",
        "ddddddddddddddddddddd"
    ]
} ----------
"""
```

### 1
Should it be relative to the enclosing node?

```python
if indent0:
    if indent1:
        if indent2:
            foo = f"""
# Heading
--------- {
                [
                    "aaaaaaaaaaaaaaaaaaaaa",
                    "bbbbbbbbbbbbbbbbbbbbb",
                    "ccccccccccccccccccccc",
                    "ddddddddddddddddddddd",
                ]
            } ----------
"""
```

### 2
Should it be relative to the opening curly brace? This is problematic because the opening curly brace could be at a position which is not in multiples of 4 as there's text before it.

```python
if indent0:
    if indent1:
        if indent2:
            foo = f"""
# heading
--------- {
      [
          "aaaaaaaaaaaaaaaaaaaaa",
          "bbbbbbbbbbbbbbbbbbbbb",
          "ccccccccccccccccccccc",
          "ddddddddddddddddddddd",
      ]
  } ----------
"""
```

### 3
Should it be always be an absolute 4 space indent from line start?

```python
if indent0:
    if indent1:
        if indent2:
            foo = f"""
# heading
--------- {
    [
        "aaaaaaaaaaaaaaaaaaaaa",
        "bbbbbbbbbbbbbbbbbbbbb",
        "ccccccccccccccccccccc",
        "ddddddddddddddddddddd",
    ]
} ----------
"""
```

Real world example: https://github.com/python-poetry/poetry/blob/f49f3ee838379c26790dda60893ff51ab431e975/tests/packages/test_locker.py#L1162-L1190

**Reply by MichaReiser** [reply]
You probably run into issues with option 3 when you have nested f-strings. Do you know how other formatters handle the indent?

**Reply by dhruvmanila** [reply]
I'm sorry but I'm unable to figure out what kind of issues are you referring to?

For example, it's always 4 space indent no matter the nesting level. The indent is relative to the start of the line.

```python
if indent0:
    if indent1:
        if indent2:
            foo = f"""
# Heading
--------- {
    f"aaaaaaa {
    [
        "aaaaaaaaaaaaaaaaaaaaa",
        "bbbbbbbbbbbbbbbbbbbbb",
        "ccccccccccccccccccccc",
        "ddddddddddddddddddddd"
    ]
} bbbbbbbb" +
    [
        "aaaaaaaaaaaaaaaaaaaaa",
        "bbbbbbbbbbbbbbbbbbbbb",
        "ccccccccccccccccccccc",
        "ddddddddddddddddddddd"
    ]
} ----------
"""
```

> Do you know how other formatters handle the indent?

I don't think any formatters implement f-string formatting ;)
Or, are you referring to any formatter outside the Python ecosystem?

**Reply by MichaReiser** [reply]
Your example demonstrates the problem. It's no longer recognizable which fstring element the } closes

I'm referring to other formatters implementing formatting of string interpolation expressions

**Reply by dhruvmanila** [reply]
Oh I see. You mean a problem in the style and not an issue in implementing the said style.

> I'm referring to other formatters implementing formatting of string interpolation expressions

[Prettier] is doing a mix of (1) and (2). Basically, it looks at closest expression indent - it could be from the parent assignment statement (`foo`) or the inner replacement field (`{`).

This actually makes sense. Currently, we do (1). The only downside of mixing (2) is that the indent is not necessarily going to be in multiples of indent width then. For example:

```python
if indent0:
    if indent1:
        if indent2:
            foo = f"""hello world
hello {
          f"aaaaaaa {
              [
                  'aaaaaaaaaaaaaaaaaaaaa',
                  'bbbbbbbbbbbbbbbbbbbbb',
                  'ccccccccccccccccccccc',
                  'ddddddddddddddddddddd'
              ]
          } bbbbbbbb"
          + [
              'aaaaaaaaaaaaaaaaaaaaa',
              'bbbbbbbbbbbbbbbbbbbbb',
              'ccccccccccccccccccccc',
              'ddddddddddddddddddddd'
          ]
      } --------
"""
```

Notice that the indentation of the inner f-string is relative to the opening curly brace on the previous line.

[Prettier]: https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAZgAgBQwE4CucAlJsADpSaYY75GkVXU1a6ElmUssCG--TAF5MAAwAWcADZSImAO4Q8UgCaZuPTVu07MAEiabRUOAGd4agxp4BtAOQDHjuwBpMdgEZfvPj6-dggUHBYP52KhGRUSphcHHxCXYAugC+opgA1OrMtg5OTmG+RWEhpWHRFbEJ1XDJ1phpANz1KRqtUO2UtOwMXMzd9JyGLAMcjPXUTsJikjJyisqWwyzGZhb6yyz2+QVunkU+JaXB5RVRVTVxyWmZ2VrbOwKFB95Hx4GnZxEXl7VJLaJmjkGm1KCkQC4QBAAA4wVDQUzIUC8PB4CDyAAKKIQiJQvCk8l4AE9EZCPHheGAANZwGAAZV4AFs4AAZVAmZDofGmOBkinU2l06GU9kAc2Qg0hPMZqAlHEhcAAHtC4HhUMzYPiACqqqAo1BmTnc3kgUxiqRwACKBAg8CNUh5kIAVqZFXTzVabXakFyHSaAI5euAYtHQ3EgXimAC0JjgKjjEJA+F4qCkYoAwhBGYzeMgIzJE2aoKKLQBBGD4VAeAjwDGqtkcn3GyHiGCMqQAdXEqHgpmFYDgdJxPdQADce0S82BTKSQKOiABJKDx2B0sBq2Gl5d0mBEi32x0gaFonkdinQvPHsyq0dwRPsnl4GAh3iinMHk3CvCPvMeXgeaRE2PdkYA7VAVBgcRkAADgABkhPA4EDVBEJfN9cybP1IRgf8wIgqCkAAFkhAgeS1f9cV9Q84EZACIjjFleGLAhXzgAAxJQcwrMU814GsIBAFIUiAA

**Reply by MichaReiser** [reply]
> Should it be relative to the opening curly brace? This is problematic because the opening curly brace could be at a position which is not in multiples of 4 as there's text before it.

This is problematic because it conflicts with lint rules? It shouldn't be a problem in the sense that it produces invalid syntax because the leading whitespace isn't an indent.

**Reply by dhruvmanila** [reply]
> This is problematic because it conflicts with lint rules?

No, I think it was mostly around visual indentation where it might get confusing to see some expressions being indented by an odd number of whitespaces. But, considering it now I think it's fine because it will be relative to the opening curly braces.

**Comment by JelleZijlstra** [comment]
Since this was brought up in https://github.com/psf/black/pull/3822#issuecomment-1948548408, here are my opinions, going over the examples in the OP:

1. Your proposed formatting seems fine.
2. Seems fine too except I don't see why you use an 8-space indent. Should be 4-space.
3. I think putting line breaks inside the expression part always looks weird and you shouldn't do it. I would want to format this as something like:
```
x = (
    f"aaaaaaaaaaaa { bbbbbbbbbbbbbb } cccccccccccccccccccc"
    f" { ddddddddddddddd } eeeeeeeeeeeeee"
)
```
However, it's hard for a formatter to decide on exact line breaks, so in Black I'd leave this up to users to do.

4. Adding whitespace on both sides looks nicer to me.
5. As above, I wouldn't break this into lines at all. I would make hugging here consistent with the preview style (i.e., hug only if you'd hug in the corresponding function call). Your second codeblock here has the `[` overindented; I assume that's a typo.
6, 7. The format specifier cannot have a newline after it, because that means the newline goes into the format spec string, and common types (e.g. `int.__format__`) will reject that newline. I think that means you have to preserve the way the user wrote the format spec. For consistency, I'd also recommend putting conversion specifiers (`!r`) immediately before the closing brace.

8. I would prefer to alternate quote styles as long as possible; it's easier for humans to read.

Self-documenting expressions section: You actually can't normalize whitespace in the format spec either, because that can affect behavior. See https://github.com/psf/black/issues/4112#issuecomment-1935250743.

**Reply by JelleZijlstra** [reply]
And one more thought: it's good that Ruff is also a linter, because most of the examples we're discussing are terrible code and a linter should probably flag them :D

**Reply by dhruvmanila** [reply]
Thanks for your valuable feedback! This is really helpful.

> 2. Seems fine too except I don't see why you use an 8-space indent. Should be 4-space.

It must be a typo. It should probably be fixed now.

> 3. I think putting line breaks inside the expression part always looks weird and you shouldn't do it. I would want to format this as something like:
> 
> ```
> x = (
>     f"aaaaaaaaaaaa { bbbbbbbbbbbbbb } cccccccccccccccccccc"
>     f" { ddddddddddddddd } eeeeeeeeeeeeee"
> )
> ```
> 
> However, it's hard for a formatter to decide on exact line breaks, so in Black I'd leave this up to users to do.

Regarding line breaks, what our current heuristic is that if any of the expression in an f-string is multiline (containing line breaks), this signals to the formatter that it can break expressions. This means that breaking expressions inside f-string is in the control of the programmer. This is similar to how Prettier does it for JavaScript's template literals (https://prettier.io/docs/en/next/rationale.html#template-literals).

> 5. As above, I wouldn't break this into lines at all. I would make hugging here consistent with the preview style (i.e., hug only if you'd hug in the corresponding function call). Your second codeblock here has the `[` overindented; I assume that's a typo.

Yes, that's a typo. I agree with the hugging point here which is yet to be implemented.

>    6, 7. The format specifier cannot have a newline after it, because that means the newline goes into the format spec string, and common types (e.g. `int.__format__`) will reject that newline. I think that means you have to preserve the way the user wrote the format spec. 

Oh I see. Thanks for noticing that. We'd need to fix it. This would basically mean that we shouldn't add any newlines in the expression part as well if the format specifier is present because that doesn't look consistent. So, the following should collapse the expression:

```python
f"""foo {
    xxxxxxx:.3f}"""
```

> For consistency, I'd also recommend putting conversion specifiers (`!r`) immediately before the closing brace.

It seems that there can't be any line breaks after the conversion flag for pre 3.12 versions even in triple-quoted strings:

```python
f"""aaaaaa {
    foo!s
} bbbbb"""
```

Which seems inconsistent because it's valid without the conversion flag:

```python
f"""aaaaaa {
    foo
} bbbbb"""
```

I need to fix this as well.

> 6. I would prefer to alternate quote styles as long as possible; it's easier for humans to read.

Yeah, I think so as well.
 
> Self-documenting expressions section: You actually can't normalize whitespace in the format spec either, because that can affect behavior. See [psf/black#4112 (comment)](https://github.com/psf/black/issues/4112#issuecomment-1935250743).

Oh I see. This is interesting.
